lib/persisted-model.js

```diff
@@ -1081,9 +1100,88 @@ PersistedModel.bulkUpdate = function(updates, callback) {
     }
   });
 
-  async.parallel(tasks, callback);
+  async.parallel(tasks, function(err) {
+    if (err) return callback(err);
+    if (conflicts.length) {
+      err = new Error('Conflict');
+      err.statusCode = 409;
+      err.details = { conflicts: conflicts };
+      return callback(err);
+    }
+    callback();
+  });
 };
 
+function applyUpdate(Model, id, data, change, conflicts, cb) {
+  var Change = Model.getChangeModel();
+  Model.findById(id, function(err, inst) {
+    if (err) return cb(err);
+
+    var rev = inst ?
+      Change.revisionForInst(inst) :
+      null;
+
+    if (rev !== change.prev) {
+      debug('Detected non-rectified change of %s %j',
+        Model.modelName, id);
+      debug('\tExpected revision: %s', change.rev);
+      debug('\tActual revision:   %s', rev);
+      conflicts.push(change);
+      return Change.rectifyModelChanges(Model.modelName, [id], cb);
+    }
+
+    // TODO(bajtos) modify `data` so that it instructs
+    // the connector to remove any properties included in "inst"
+    // but not included in `data`
+    // See https://github.com/strongloop/loopback/issues/1215
+
+    Model.updateAll(inst.toObject(), data, function(err, count) {
+      if (err) return cb(err);
+      if (count === 0) {
+        debug('UpdateAll detected non-rectified change of %s %j',
+          Model.modelName, id);
+        conflicts.push(change);
+        // NOTE(bajtos) updateAll triggers change rectification
+        // for all model instances, even when no records were updated,
+        // thus we don't need to rectify explicitly ourselves
+      } else if (count !== 1) {
+        debug('%s.updateAll modified unexpected number of instances: %j',
+          Model.modelName, count);
+        return cb(new Error(
```

Comment:
`count !== 1` is really a good enough check for this? If `count` is `2` we should report an error? What about `typeof count` or `arguments.length`...